PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -534,6 +743,259 @@ void ProjectPanel::openSelectFile()
+			break;
+		case nodeType_project:
+			iImg = INDEX_PROJECT;
+			break;
+		case nodeType_folder:
+			iImg = (tvItem.state & TVIS_EXPANDED) ? INDEX_OPEN_NODE : INDEX_CLOSED_NODE;
+			break;
+		case nodeType_file:
+			iImg = ::PathFileExists(data._filePath.c_str()) ? INDEX_LEAF : INDEX_LEAF_INVALID;
+			break;
+		case nodeType_dirFile:
+			iImg = INDEX_LEAF_DIRFILE;
+			break;
+		case nodeType_baseDir:
+		case nodeType_dir:
+			rebuildDirectoryTree(hTreeItem, data);
+			return;
+		default:
+			return;
+	}
+	_treeView.setItemImage(hTreeItem, iImg, iImg);
+
+}
+
+generic_string ProjectPanel::buildDirectoryName(const generic_string& filePath, const std::vector<generic_string>& filters)
+{
+	generic_string result(filePath);
+
+	const size_t lastSlashIdx = result.find_last_of(TEXT("\\/"));
+	if (std::string::npos != lastSlashIdx)
+	{
+		result.erase(0, lastSlashIdx + 1);
+		if( result.empty() ) // drive
+		{
+			result = filePath;
+			result.erase(lastSlashIdx);
+		}
+	}
+
+	if (!filters.empty())
+	{
+		result += TEXT(" (");
+		result += stringJoin(filters, TEXT(";"), true);
+		result += TEXT(")");
+	}
+
+	return result;
+
+}
+
+void ProjectPanel::itemVisibilityChanges(HTREEITEM hItem, bool visible)
+{
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	bool expanded = (tvItem.state & TVIS_EXPANDED) != 0;
+	if (data.isDirectory())
+	{
+		if (visible && expanded)
+			data.watchDir(true);
+		if (!visible || !expanded)
+			data.watchDir(false);
+	}
+
+	for (HTREEITEM hChildItem = _treeView.getChildFrom(hItem);
+    hChildItem != NULL;
+    hChildItem = _treeView.getNextSibling(hChildItem))
+    {
+		itemVisibilityChanges(hChildItem, visible);
+	}
+
+}
+
+void ProjectPanel::expandOrCollapseDirectory(bool expand, HTREEITEM hItem)
+{
+
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	if (!(data.isDirectory() || data.isBaseDirectory()) )
+		return;
+
+	if (expand)
+	{
+		removeDummies(hItem);
+		if (data.isDirectory())
+			data.watchDir(true);
+	}
+	else
+	{
+		if (data.isDirectory())
+			data.watchDir(false);
+	}
+
+	for (HTREEITEM hChildItem = _treeView.getChildFrom(hItem);
+    hChildItem != NULL;
+    hChildItem = _treeView.getNextSibling(hChildItem))
+    {
+		itemVisibilityChanges(hChildItem, expand);
+	}
+
+	_directoryWatcher.update();
+
+}
+
+bool ProjectPanel::setFilters(const std::vector<generic_string>& filters, HTREEITEM hItem)
+{
+	TVITEM tvItem;
+	tvItem.mask = TVIF_PARAM;
+	tvItem.hItem = hItem;
+	::SendMessage(_treeView.getHSelf(), TVM_GETITEM, 0, reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.